### PR TITLE
feature/Box-addon

### PR DESCRIPTION
**Pull request by mfraezz**
Purpose
=======
Add Box to the supported addons
Accompanies: https://github.com/CenterForOpenScience/waterbutler/pull/9

Changes
=======
Users can now connect with Box as an OSF addon.

Side Effects
=========
See the [README](https://github.com/mfraezz/osf.io/blob/develop/website/addons/box/README.md) for setup instructions

### CODE REVIEW

File: website/addons/box/client.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    """
+    user_settings = user.get_addon('box')
+    if not user_settings:
+        raise AddonError('User does not have the Box addon enabled.')
+    if refresh_creds_if_necessary(user_settings):
+        return BoxClient(user_settings.get_credentialsv2())
+    raise AddonError('Box credentials for this user have expired.')
+
+
+def get_client_from_user_settings(settings_obj):
```
Comment (chrisseto) [inline website/addons/box/client.py]: Shouldn't these methods be reversed? `get_client(user)` should call `get_client_from_user_settings`

File: website/addons/box/model.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+        if add_log:
+            extra = {'folder': folder}
+            nodelogger = BoxNodeLogger(node=node, auth=auth)
+            nodelogger.log(action="node_deauthorized", extra=extra, save=True)
+
+    def serialize_waterbutler_credentials(self):
+        if not self.has_auth:
+            raise exceptions.AddonError('Addon is not authorized')
+        return {'token': self.user_settings.access_token}
```
Comment (chrisseto) [inline website/addons/box/model.py]: Should you attempt to update the access token here?

File: website/addons/box/model.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+                'path': cleaned_path,
+                'folder': self.folder,
+                'urls': {
+                    'view': self.owner.web_url_for('box_view_file', path=cleaned_path),
+                    'download': self.owner.web_url_for('box_download', path=cleaned_path),
+                },
+            },
+        )
+
+    def get_waterbutler_render_url(self, path, rev=None, **kwargs):
```
Comment (chrisseto) [inline website/addons/box/model.py]: This method is deprecated as of 0.25, go ahead and remove it.

File: website/addons/box/model.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        self.owner.add_log(
+            'box_{0}'.format(action),
+            auth=auth,
+            params={
+                'project': self.owner.parent_id,
+                'node': self.owner._id,
+                'path': cleaned_path,
+                'folder': self.folder,
+                'urls': {
+                    'view': self.owner.web_url_for('box_view_file', path=cleaned_path),
```
Comment (chrisseto) [inline website/addons/box/model.py]: In as of 0.25 you can use `self.owner.web_url_for('addon_download_or_view_file', provider='box', action='view/download', path=...)`

File: website/addons/box/routes.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            ],
+            'get',
+            views.auth.box_oauth_start,
+            json_renderer,
+        ),
+    ],
+    'prefix': '/api/v1'
+}
+
+web_routes = {
```
Comment (chrisseto) [inline website/addons/box/routes.py]: These can be removed.

File: website/addons/box/routes.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+        Rule(
+            ['/project/<pid>/box/config/share/',
+            '/project/<pid>/node/<nid>/box/config/share/'],
+            'get',
+            views.config.box_get_share_emails,
+            json_renderer
+        ),
+
+        ##### CRUD #####
```
Comment (chrisseto) [inline website/addons/box/routes.py]: Crud should be defined in waterbutler not in OSF

File: website/addons/box/routes.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                '/project/<pid>/box/files/<path:path>',
+                '/project/<pid>/node/<nid>/box/files/',
+                '/project/<pid>/node/<nid>/box/files/<path:path>',
+            ],
+            'post',
+            views.crud.box_upload,
+            json_renderer
+        ),
+
+        ##### File rendering #####
```
Comment (chrisseto) [inline website/addons/box/routes.py]: File rendering and file revisions are taken care of by 0.25 and waterbutler, respectively.

File: website/addons/box/static/loading-bars.svg [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32" width="32" height="32" fill="blue">
```
Comment (chrisseto) [inline website/addons/box/static/loading-bars.svg]: What is this file?

File: website/addons/box/static/file-detail.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+var RevisionTable = require('./revisions.js');
```
Comment (chrisseto) [inline website/addons/box/static/file-detail.js]: This can be removed.

File: website/addons/box/static/revisions.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
+/**
```
Comment (chrisseto) [inline website/addons/box/static/revisions.js]: This file can also be removed.

File: website/addons/box/model.py (hunk 2 above)
Comment (mfraezz) [inline website/addons/box/model.py]: It is being attempted here. Any time has_auth is called it will refresh the tokens if needed.

File: website/addons/box/model.py [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+class BoxFile(GuidFile):
+    """A Box file model with a GUID. Created lazily upon viewing a
+    file's detail page.
+    """
+
+    #: Full path to the file, e.g. 'My Pictures/foo.png'
+    path = fields.StringField(required=True, index=True)
+
+    @property
+    def file_name(self):
```
Comment (chrisseto) [inline website/addons/box/model.py]: Dont override this, its dropbox specific, the default should work fine.

File: website/addons/box/model.py [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+    @property
+    def provider(self):
+        return 'box'
+
+    @property
+    def version_identifier(self):
+        return 'revision'
+
+    @property
+    def unique_identifier(self):
```
Comment (chrisseto) [inline website/addons/box/model.py]: If you add `etag` to revisions add it here otherwise the revision should work just fine

File: website/addons/box/model.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+        return self.path.split('/')[1][-5] #self._metadata_cache['extra']['revisionId']
+
+    @classmethod
+    def get_or_create(cls, node, path):
+        """Get or create a new file record. Return a tuple of the form (obj, created)
+        """
+        cleaned_path = clean_path(path)
```
Comment (chrisseto) [inline website/addons/box/model.py]: You shouldn't need to clean the path

File: website/addons/box/model.py [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+
+    @property
+    def waterbutler_path(self):
+        path = self.path
+        if self.folder == '/':
+            return path
+        return path.replace(self.folder, '', 1)
+
+    @property
+    def folder(self):
```
Comment (chrisseto) [inline website/addons/box/model.py]: Not required for box

File: website/addons/box/model.py [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+
+    def token_refreshed_callback(self, access_token, refresh_token):
+        self.access_token = access_token
+        self.refresh_token = refresh_token
+        self.last_refreshed = datetime.utcnow()
+        self.save()
+
+    @property
+    def has_auth(self):
+        return bool(self.access_token and refresh_creds_if_necessary(self))
```
Comment (chrisseto) [inline website/addons/box/model.py]: Can `refresh_creds_if_necessary` be moved into this class as self is being passed to it anyways?

File: website/addons/box/model.py [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+        return u'<BoxUserSettings(user={self.owner.username!r})>'.format(self=self)
+
+
+class BoxNodeSettings(AddonNodeSettingsBase):
+
+    user_settings = fields.ForeignField(
+        'boxusersettings', backref='authorized'
+    )
+
+    folder = fields.StringField(default=None)
```
Comment (chrisseto) [inline website/addons/box/model.py]: Could the name of a folder and not the id change? If so we shouldn't be caching that property.

File: website/addons/box/model.py [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+    folder_id = fields.IntegerField(default=None)
+
+    #: Information saved at the time of registration
+    #: Note: This is unused right now
+    registration_data = fields.DictionaryField()
```
Comment (chrisseto) [inline website/addons/box/model.py]: Not required, remove this field

File: website/addons/box/model.py [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+
+
+    @property
+    def display_name(self):
+        return '{0}: {1}'.format(self.config.full_name, self.folder)
+
+    @property
+    def has_auth(self):
+        """Whether an access token is associated with this node."""
+        return bool(self.user_settings and self.user_settings.has_auth and refresh_creds_if_necessary(self.user_settings))
```
Comment (chrisseto) [inline website/addons/box/model.py]: `user_settings.has_auth` will already call refresh if necessary

File: website/addons/box/model.py [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+        """Import a user's Box authentication and create a NodeLog.
+
+        :param BoxUserSettings user_settings: The user settings to link.
+        """
+        self.user_settings = user_settings
+        nodelogger = BoxNodeLogger(node=self.owner, auth=Auth(user_settings.owner))
+        nodelogger.log(action="node_authorized", save=True)
+
+    def find_or_create_file_guid(self, path):
+        return BoxFile.get_or_create(self.owner, clean_path(os.path.join(self.folder, path.lstrip('/'))))
```
Comment (chrisseto) [inline website/addons/box/model.py]: As box is a new addon you can just pass the path in here without any transformation.
The only reason path is mutated else where is to avoid have to do any form of migration.

File: website/addons/box/settings/local-dist.py [hunk 20]
```diff
@@ -0,0 +1,7 @@
+# -*- coding: utf-8 -*-
+"""Example Box local settings file. Copy this file to local.py and change
+these settings.
+"""
+# Get an app key and secret at https://www.box.com/developers/apps
+BOX_KEY = 'changeme'
```
Comment (chrisseto) [inline website/addons/box/settings/local-dist.py]: Maybe set these to `None` so there wont be any obscure error messages.

File: website/addons/box/utils.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+    abs_path = os.path.abspath(path).lower()
+    return os.path.commonprefix([abs_path, abs_directory]) == abs_directory
+
+
+def is_authorizer(auth, node_addon):
+    """Return if the auth object's user is the same as the authorizer of the node."""
+    return auth.user == node_addon.user_settings.owner
+
+
+def abort_if_not_subdir(path, directory):
```
Comment (chrisseto) [inline website/addons/box/utils.py]: This function may not be relevant to Box and it probably wont work either.

File: website/addons/box/utils.py [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+
+
+def get_file_name(path):
+    """Given a path, get just the base filename.
+    Handles "/foo/bar/baz.txt/" -> "baz.txt"
+    """
+    return os.path.basename(path.strip('/'))
+
+
+def clean_path(path):
```
Comment (chrisseto) [inline website/addons/box/utils.py]: Box should not require this function.

File: website/addons/box/utils.py [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+    """Ensure a path is formatted correctly for url_for."""
+    if path is None:
+        return ''
+    return path.strip('/')
+
+
+def make_file_response(fileobject, metadata):
```
Comment (chrisseto) [inline website/addons/box/utils.py]: This is no longer used.

File: website/addons/box/utils.py [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+    if not path.startswith('/'):
+        return '/' + path
+    return path
+
+
+def build_box_urls(item, node):
+    path = clean_path(item['path'])  # Strip trailing and leading slashes
+    if item['type']==u'folder':
+        return {
+            'upload': node.api_url_for('box_upload', path=path),
```
Comment (chrisseto) [inline website/addons/box/utils.py]: This end point doesnt exist

File: website/addons/box/utils.py [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+            # Endpoint for fetching all of a folder's contents
+            'fetch': node.api_url_for('box_hgrid_data_contents', path=path),
+            # Add extra endpoint for fetching folders only (used by node settings page)
+            # NOTE: querystring params in camel-case
+            'folders': node.api_url_for('box_hgrid_data_contents',
+                path=path, foldersOnly=1)
+        }
+    else:
+        return {
```
Comment (chrisseto) [inline website/addons/box/utils.py]: This should never be run, remove it and for safety add at the top of the function:
```python
assert item['type'] == 'folder', 'Can only build urls for box folders'
```

File: website/addons/box/utils.py [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+        'ext': os.path.splitext(filename)[-1],
+        rubeus.KIND: rubeus.FOLDER if item['type']==u'folder' else rubeus.FILE,
+        #'urls': build_box_urls(item, node),
+        'path': item['name'],
+        'id': item['id']
+    }
+    return serialized
+
+
+def get_share_folder_uri(path):
```
Comment (chrisseto) [inline website/addons/box/utils.py]: Is this relevant to Box?

File: website/addons/box/model.py (hunk 12 above)
Comment (mfraezz) [inline website/addons/box/model.py]: `etag` is only supported relating to versioning if you want to delete a specific old version of a file, and even then it's optional; to prevent race conditions. Is this desired functionality?

Comment (chrisseto) [inline website/addons/box/model.py]: [Etags](https://en.wikipedia.org/wiki/HTTP_ETag) are fantastic for `unique_identifier`. Take a look at S3's fileguid
